File: solutions/22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:

    vector<string> ans;

    void rec(int level, int cnt, string str,int n){
        cout<<str<<endl;
        if(level>2*n || cnt<0 || cnt>n){
            return;
        }
        if(level==2*n){
            if(cnt==0){
                ans.push_back(str);
            }else{
                return;
            }
        }

        str+='(';
        rec(level+1,cnt+1,str,n);
        str.pop_back();
        str+=')';
        rec(level+1,cnt-1,str,n);
        str.pop_back();
    }

    vector<string> generateParenthesis(int n) {
        int cnt=0;
        string str="";
        rec(0,cnt,str,n);
        return ans;
    }
};

File: solutions/22-generate-parentheses/generate-parentheses.java
class Solution {
    List<String> ans=new ArrayList<String>();
    StringBuilder s=new StringBuilder();
    public void rec(int c1,int c2){
        if(c1==0 && c2==0){
            ans.add(s.toString());
            return;
        }
        if(c1>0){
            s.append('(');
            rec(c1-1,c2);
            s.deleteCharAt(s.length()-1);
        }
        if(c1<c2){
            s.append(')');
            rec(c1,c2-1);
            s.deleteCharAt(s.length()-1);
        }
    }

    public List<String> generateParenthesis(int n) {
        int c1=n;
        int c2=n;
        rec(c1,c2);
        return ans;
    }
}



// below here s+'(' will not update s
/*
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();

        dfs(0, 0, "", n, res);

        return res;        
    }

    private void dfs(int openP, int closeP, String s, int n, List<String> res) {
        if (openP == closeP && openP + closeP == n * 2) {
            res.add(s);
            return;
        }

        if (openP < n) {
            dfs(openP + 1, closeP, s + "(", n, res);
        }

        if (closeP < openP) {
            dfs(openP, closeP + 1, s + ")", n, res);
        }
    }    
}
*/